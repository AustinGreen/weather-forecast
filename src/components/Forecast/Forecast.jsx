/* eslint-disable no-mixed-operators */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import moment from 'moment';
import icons from '../../shared/Icons';
import TransparentBackground from '../../shared/Backgrounds';
import {
  PATH_BASE,
  PARAM_TYPE,
  PARAM_UNIT,
  PARAM_APPID,
  PATH_FORECAST,
  PARAM_DAYS,
} from '../../constants';

const Loader = styled.a`
  font-size: 3rem;
  border: 0;
`;

const ImageContainer = styled.figure`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 2em auto 1em;
`;

const Card = styled.div`
  transition: all 0.2s ease-in-out;
  background-color: rgba(255,255,255,0.7);
  &:hover {
    box-shadow: 0 4px 3px rgba(10, 10, 10, 0.1), 0 0 0 1px rgba(10, 10, 10, 0.1);
    transform: scale(1.1);
  }
`;

const Days = ({ forecastResult, match }) => (
  <div className="columns">
    {forecastResult.list.map((days, i) => {
      const dateObject = new Date(days.dt * 1000);
      const formattedDate = moment(dateObject).format('ddd[,] MMMM Do');
      return (
        <div key={days.dt} className="column">
          <Link
            to={{
              pathname: `${match.url}/details`,
              hash: `#${i}`,
              state: { forecastResult },
            }}
          >
            <Card className="card">
              <div className="card-content">
                <p className="subtitle is-4">{formattedDate}</p>
                <p className="title is-2">{Math.round(days.temp.day)} ÂºF</p>
                <ImageContainer className="image is-128x128">
                  <img src={icons[`icon${days.weather[0].icon}`]} alt="" />
                </ImageContainer>
              </div>
            </Card>
          </Link>
        </div>
      );
    })}
  </div>
);

class Forecast extends Component {
  constructor(props) {
    super(props);
    this.state = {
      forecastResult: '',
      isLoading: true,
    };

    this.updateResult = this.updateResult.bind(this);
  }

  componentWillMount() {
    if (this.props.history.location.state) {
      this.updateResult(this.props.history.location.state);
    } else {
      fetch(
        `${PATH_BASE}${PATH_FORECAST}${this.props.match.params.city}&${PARAM_TYPE}&${PARAM_UNIT}&${PARAM_APPID}&${PARAM_DAYS}`,
      )
        .then(response => response.json())
        .then(forecastResult => this.updateResult(forecastResult));
    }
  }

  updateResult(forecastResult) {
    this.setState({
      forecastResult,
      isLoading: false,
    });
  }

  render() {
    const { forecastResult, isLoading } = this.state;
    const { match } = this.props;
    return (
      <TransparentBackground
        className="section column is-10
  is-offset-1 has-text-centered" large
      >
        {isLoading
          ? <Loader className="button is-loading">
              Button
            </Loader>
          : <div>
            <h1 className="title">{forecastResult.city.name} Forecast</h1>
            <Days match={match} forecastResult={forecastResult} />
          </div>}
      </TransparentBackground>
    );
  }
}

Forecast.propTypes = {
  match: PropTypes.shape({
    url: PropTypes.string,
    params: PropTypes.object,
  }).isRequired,
  history: PropTypes.shape({
    location: PropTypes.object,
    state: PropTypes.object,
  }).isRequired,
};

Days.propTypes = {
  match: PropTypes.shape({
    url: PropTypes.string,
  }).isRequired,
  forecastResult: PropTypes.shape({}).isRequired,
};

export default Forecast;
