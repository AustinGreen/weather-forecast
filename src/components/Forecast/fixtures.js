const forecastTestData = {
  city: {
    id: 2643743,
    name: 'London',
    coord: { lon: -0.1258, lat: 51.5085 },
    country: 'GB',
    population: 0,
  },
  cod: '200',
  message: 0.6052046,
  cnt: 5,
  list: [
    {
      dt: 1492513200,
      temp: {
        day: 279.41,
        min: 275.02,
        max: 279.41,
        night: 275.02,
        eve: 279.41,
        morn: 279.41,
      },
      pressure: 1038.49,
      humidity: 50,
      weather: [{ id: 801, main: 'Clouds', description: 'few clouds', icon: '02d' }],
      speed: 4.44,
      deg: 43,
      clouds: 20,
    },
    {
      dt: 1492599600,
      temp: {
        day: 284.14,
        min: 274.36,
        max: 285.02,
        night: 278.6,
        eve: 284.32,
        morn: 274.36,
      },
      pressure: 1044.05,
      humidity: 57,
      weather: [{ id: 800, main: 'Clear', description: 'sky is clear', icon: '02d' }],
      speed: 1.86,
      deg: 88,
      clouds: 8,
    },
    {
      dt: 1492686000,
      temp: {
        day: 285.8,
        min: 278.55,
        max: 288.1,
        night: 282.32,
        eve: 287.72,
        morn: 278.55,
      },
      pressure: 1042.28,
      humidity: 64,
      weather: [{ id: 500, main: 'Rain', description: 'light rain', icon: '10d' }],
      speed: 3.07,
      deg: 264,
      clouds: 8,
      rain: 0.71,
    },
    {
      dt: 1492772400,
      temp: {
        day: 288.5,
        min: 279.62,
        max: 289.79,
        night: 283.56,
        eve: 289.19,
        morn: 279.62,
      },
      pressure: 1039.75,
      humidity: 66,
      weather: [{ id: 803, main: 'Clouds', description: 'broken clouds', icon: '04d' }],
      speed: 3.86,
      deg: 305,
      clouds: 56,
    },
    {
      dt: 1492858800,
      temp: {
        day: 285.81,
        min: 278.83,
        max: 286.41,
        night: 278.83,
        eve: 286.41,
        morn: 280.03,
      },
      pressure: 1036.02,
      humidity: 0,
      weather: [{ id: 500, main: 'Rain', description: 'light rain', icon: '10d' }],
      speed: 5.34,
      deg: 320,
      clouds: 32,
      rain: 0.39,
    },
  ],
};

export default forecastTestData;
