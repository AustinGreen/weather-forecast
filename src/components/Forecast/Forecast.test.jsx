/* eslint-disable */

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Forecast from './Forecast';
import { shallow } from 'enzyme';
import forecastTestData from './fixtures';

describe('Forecast', () => {
  it('renders', () => {
    const div = document.createElement('div');
    const forecastResult = forecastTestData;
    ReactDOM.render(
      <Forecast
        history={{
          location: {
            pathname: '/forecast/london,uk',
            state: forecastResult,
          },
        }}
        match={{ params: { city: 'london,uk' } }}
        location={{
          state: { forecastResult },
        }}
      />,
      div,
    );
  });

  test('snapshots', () => {
    const forecastResult = forecastTestData;
    const component = renderer.create(
      <Forecast
        history={{
          location: {
            pathname: '/forecast/london,uk',
            state: forecastResult,
          },
        }}
        match={{ params: { city: 'london,uk' } }}
        location={{
          state: { forecastResult },
        }}
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
