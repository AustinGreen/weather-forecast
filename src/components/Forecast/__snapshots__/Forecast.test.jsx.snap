exports[`Forecast snapshots 1`] = `
<div>
  <div
    className="section column is-half is-offset-one-quarter is-large has-text-centered">
    <a
      className="button is-loading gLszEl">
      Button
    </a>
  </div>
</div>
`;
